File: full_case_study_solns/_d_mass/python_old/hw8/massController.py
import numpy as np
import massParamHW8 as P
import sys
sys.path.append('..')  # add parent directory
import massParam as P0
from PDControl import PDControl


class massController:

    def __init__(self):
        # Instantiates the PD object
        self.zCtrl = PDControl(P.kp, P.kd, P0.F_max, P0.beta, P0.Ts)
        self.limit = P0.F_max

    def u(self, y_r, y):
        # y_r is the referenced input
        # y is the current state
        z_r = y_r[0]
        z = y[0]

        tau_tilde = self.zCtrl.PD(z_r, z, False)
        tau = self.saturate(tau_tilde)
        return [tau]

    def saturate(self, u):
        if abs(u) > self.limit:
            u = self.limit*np.sign(u)
        return u

File: full_case_study_solns/_d_mass/matlab_old/hw7/massController.m
classdef massController
    % 
    %    This class inherits other controllers in order to organize multiple controllers.
    %
    %----------------------------
    properties
        zCtrl
        limit
    end
    %----------------------------
    methods
        %----------------------------
        function self = massController(P)
            self.zCtrl = PDControl(P.kp, P.kd, P.F_max, P.beta, P.Ts);
            % plant parameters known to controller
            self.limit = P.F_max;
        end
        %----------------------------
        function force = u(self, y_r, y)
            % y_r is the referenced input
            % y is the current state
            z_r = y_r;
            z = y(1);
            
            % compute the linearized torque using PID
            force_tilde = self.zCtrl.PD(z_r, z, false);
            % compute total torque
            force = self.saturate(force_tilde);
        end
        %----------------------------
        function out = saturate(self,u)
            if abs(u) > self.limit
                u = self.limit*sign(u);
            end
            out = u;
        end        
    end
end